src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/internal/utils/BufferedSourceExtensions.kt

```diff
@@ -0,0 +1,110 @@
+package it.krzeminski.snakeyaml.engine.kmp.internal.utils
+
+import okio.BufferedSource
+
+/**
+ * @return a well-formed (if possible) UTF8 string with a size at most [limitBytes]
+ */
+// based on benchmarks it looks like having inline here actually makes a small difference in the performance
+@Suppress("NOTHING_TO_INLINE")
+internal inline fun BufferedSource.readUtf8WithLimit(limitBytes: Long): String {
+    val fullUtf8Size = sizeOfFullValidUtf8String(limitBytes)
+    val buffer = readUtf8(fullUtf8Size)
+    return buffer
+}
+
+@Suppress("NOTHING_TO_INLINE")
+// based on benchmarks it looks like having inline here actually makes a small difference in the performance
+private inline fun BufferedSource.sizeOfFullValidUtf8String(limitBytes: Long): Long {
+    val readBytes = limitBytes
+    val hasAll = request(readBytes)
+
+    val originalSize = if (hasAll) readBytes else buffer.size
```

Comment:
nit: probably no need to introduce another val:
```suggestion
    val hasAll = request(limitBytes)

    val originalSize = if (hasAll) limitBytes else buffer.size
```

Reply:
Thanks for spotting. Just forgot to remove after refactoring I did